website/blog/2022-07-12-change-data-capture-metrics.md

```diff
@@ -0,0 +1,348 @@
+}}
+```
+
+As a final step, Joanne would create `fct_income_history` to join in the seed file to determine which versions are “correct”. Her new DAG looks like this, where `int_income_history` is an incremental model without a unique key:
+
+![](/img/blog/2022-07-12-change-data-capture-metrics/int-income-history-dag.png)
+
+The final output of `fct_income_history` would look identical to `stg_snapshot_fct_income` from her initial approach: 
+
+| month_year | product_category | revenue | run_timestamp | correct_version |
+|:---:|:---:|:---:|:---:|:---:|
+| January 2022 | clothing | 100 | 01/30/22 12:00:00 | FALSE |
+| January 2022 | electronics | 200 | 01/30/22 12:00:00 | FALSE |
+| January 2022 | books | 300 | 01/30/22 12:00:00 | FALSE |
+| January 2022 | clothing | 50 | 02/03/22 16:00:00 | FALSE |
+| January 2022 | electronics | 150 | 02/03/22 16:00:00 | FALSE |
+| January 2022 | books | 200 | 02/03/22 16:00:00 | FALSE |
+| January 2022 | clothing | 52 | 02/10/22 08:00:00 | TRUE |
+| January 2022 | electronics | 152 | 02/10/22 08:00:00 | TRUE |
+| January 2022 | books | 202 | 02/10/22 08:00:00 | TRUE |
+
+## Solution #2: Upstream Snapshots
+
+Alternatively, Joanne could snapshot her source data and add flexibility to her modeling so that all historical versions are calculated `at the same time`. Let’s look at our example.
+
+Joanne could track changes in the source data by adding snapshots directly on top of her raw data.
+
+![](/img/blog/2022-07-12-change-data-capture-metrics/snapshots-dag.png)
+
+This would *change the <Term id="grain" />* of these `stg_` tables, so she would see a row for each version of each field. The staging models will contain the history of each record. 
+
+Remember the source data change Joanne noticed — a manufacturing cost was dated incorrectly (Junkuary 2022 instead of January 2022). With this solution, the `costs_snapshot` model will pick up this change:
+
+```sql
+{% snapshot costs_snapshot %}
+
+{{
+    config(
+      target_database='analytics',
+      target_schema='snapshots',
+      unique_key='id',
+
+      strategy='timestamp',
+      updated_at='updated_at',
+    )
+}}
+
+select * from {{ source('source', 'costs') }}
+
+{% endsnapshot %}
+```
+
+| id | month_year | cost | updated_at | dbt_valid_from | dbt_valid_to |
+|:---:|:---:|:---:|:---:|:---:|:---:|
+| 1 | Junkuary 2022 | 50 | 01/15/22 12:00:00 | 01/15/22 12:00:00 | 02/03/22 12:00:00 |
+| 1 | January 2022 | 50 | 02/03/22 12:00:00 | 02/03/22 12:00:00 | NULL |
+
+:::note Note
+Because snapshots only capture changes detected at the time the dbt snapshot command is executed, it is technically possible to miss some changes to your source data. You will have to consider how often you want to run this snapshot command in order to capture the history you need.
+:::
+
+`fct_income_history` now calculates the metrics for each version of source data, every time Joanne executes a `dbt run`. In other words, the downstream `fct_` models are **version-aware**.
```

Comment:
```suggestion
The original `fct_income` model now calculates the metrics for each version of source data, every time Joanne executes a `dbt run`. In other words, the downstream `fct_` models are **version-aware**. Because of this, Joanne changes the name of `fct_income` to `fct_income_history` to be more descriptive.
```